frontend/src/liboperator/operator/devices.go

```diff
@@ -138,3 +211,28 @@ func (p *DevicePool) DeviceIds() []string {
 	}
 	return ret
 }
+
+func (p *DevicePool) GetDeviceInfoList() []*DeviceInfo {
+	p.devicesMtx.Lock()
+	defer p.devicesMtx.Unlock()
+	ret := make([]*DeviceInfo, 0)
+	for id, device := range p.devices {
+		ret = append(ret, &DeviceInfo{DeviceId: id, GroupId: device.groupId})
+	}
+	return ret
+}
+
+func (p *DevicePool) GetDeviceInfoListByGroupId(groupId string) []*DeviceInfo {
```

Comment:
can we add GetDeviceInfoListByPredicate or something, and let GetDeviceInfoList and GetDeviceInfoListByGroupId use it?

Reply:
To do that I think GetDeviceInfoList should be first refactored to refer group map of the pool(p.group) to get all devices:

```
func (p *DevicePool) GetDeviceInfoList() []*DeviceInfo {
	p.devicesMtx.Lock()
	defer p.devicesMtx.Unlock()
	ret := make([]*DeviceInfo, 0)

	for groupId, group := range p.groups {
		for _, deviceId := range group.deviceIds {
			ret = append(ret, &DeviceInfo{DeviceId: deviceId, GroupId: groupId})
		}
	}
	return ret
}
```

The only common part between GetDeviceInfoList and GetDeviceInfoListByGroupId becomes the following snippet,
thus GetDeviceInfoListByPredicate do the following:
```
	for _, deviceId := range group.deviceIds {
		ret = append(ret, &DeviceInfo{DeviceId: deviceId, GroupId: groupId})
	}
```

Is this what you mean by?

Reply:
nvm, I just thought we could get deviceIds from p.devices with predicates, but for group, the current code looks better.